include/zephyr/drivers/gpio.h

```diff
@@ -417,6 +417,252 @@ struct gpio_dt_spec {
+ *        and @p "ngpios" DT properties values
+ *
+ * @param node_id GPIO controller node identifier.
+ * @return the bitmask of reserved gpios
+ */
+#define GPIO_DT_RESERVED_RANGES(node_id)				       \
+	GPIO_DT_RESERVED_RANGES_NGPIOS(node_id, DT_PROP(node_id, ngpios))
+
+/**
+ * @brief Makes a bitmask of reserved GPIOs from a DT_DRV_COMPAT instance's
+ *        @p "gpio-reserved-ranges" property and @p "ngpios" argument
+ *
+ * @param inst DT_DRV_COMPAT instance number
+ * @return the bitmask of reserved gpios
+ * @param ngpios  number of GPIOs
+ * @see GPIO_DT_RESERVED_RANGES()
+ */
+#define GPIO_DT_INST_RESERVED_RANGES_NGPIOS(inst, ngpios)		       \
+		GPIO_DT_RESERVED_RANGES_NGPIOS(DT_DRV_INST(inst), ngpios)
+
+/**
+ * @brief Make a bitmask of reserved GPIOs from a DT_DRV_COMPAT instance's GPIO
+ *        @p "gpio-reserved-ranges" and @p "ngpios" properties
+ *
+ * @param inst DT_DRV_COMPAT instance number
+ * @return the bitmask of reserved gpios
+ * @see GPIO_DT_RESERVED_RANGES()
+ */
+#define GPIO_DT_INST_RESERVED_RANGES(inst)				       \
+		GPIO_DT_RESERVED_RANGES(DT_DRV_INST(inst))
+
+/**
+ * @brief Makes a bitmask of allowed GPIOs from DT @p "gpio-reserved-ranges"
+ *        property and @p "ngpios" argument
+ *
+ * This macro is paired with GPIO_DT_RESERVED_RANGES_NGPIOS(), however unlike
+ * the latter, it returns a bitmask of ALLOWED gpios.
+ *
+ * Example devicetree fragment:
+ *
+ * @code{.dts}
+ *	a {
+ *		compatible = "some,gpio-controller";
+ *		ngpios = <32>;
+ *		gpio-reserved-ranges = <0 8>, <9 5>, <15 16>;
+ *	};
+ *
+ * @endcode
+ *
+ * Example usage:
+ *
+ * @code{.c}
+ *	struct some_config {
+ *		uint32_t port_pin_mask;
+ *	};
+ *
+ *	static const struct some_config dev_cfg = {
+ *		.port_pin_mask = GPIO_DT_PORT_PIN_MASK_NGPIOS_EXC(
+ *					DT_LABEL(a), 32),
+ *	};
+ * @endcode
+ *
+ * This expands to:
+ *
+ * @code{.c}
+ *	struct some_config {
+ *		uint32_t port_pin_mask;
+ *	};
+ *
+ *	static const struct some_config dev_cfg = {
+ *		.port_pin_mask = 0x80004100,
+ *				// 0b1000 0000 0000 0000 0100 0001 00000 000
+ *	};
+ * @endcode
+ *
+ * @param node_id GPIO controller node identifier.
+ * @param ngpios  number of GPIOs
+ * @return the bitmask of allowed gpios
+ */
+#define GPIO_DT_PORT_PIN_MASK_NGPIOS_EXC(node_id, ngpios)		       \
+	((gpio_port_pins_t)						       \
+	COND_CODE_0(ngpios,						       \
+		(0),							       \
+		(COND_CODE_1(DT_NODE_HAS_PROP(node_id, gpio_reserved_ranges),  \
+			((GENMASK64(ngpios - 1, 0) &			       \
+			~GPIO_DT_RESERVED_RANGES_NGPIOS(node_id, ngpios))),    \
+			(GENMASK64(ngpios - 1, 0)))			       \
+		)							       \
+	))
+
+/**
+ * @brief Makes a bitmask of allowed GPIOs from the @p "gpio-reserved-ranges"
+ *        and @p "ngpios" DT properties values
+ *
+ * @param node_id GPIO controller node identifier.
+ * @return the bitmask of allowed gpios
+ */
+#define GPIO_DT_PORT_PIN_MASK_EXC(node_id)				       \
```

Comment:
Why not roll this into the existing `GPIO_PORT_PIN_MASK_FROM_NGPIOS()`? It would seem with the current approach, each driver must be modified in order to support reserved GPIOs?

Reply:
I'm supposing to do this as a separate PR: exactly because all gpios drivers will be affected (and some unknown external modules too), so it can be revert back in case of a regression. The current PR does not affect anything except our code for now.

Reply:
I mean this one:
` #define GPIO_PORT_PIN_MASK_FROM_NGPIOS()           GPIO_DT_PORT_PIN_MASK_NGPIOS_EXC()`

Reply:
> I'm supposing to do this as a separate PR: exactly because all gpios drivers will be affected (and some unknown external modules too), so it can be revert back in case of a regression. The current PR does not affect anything except our code for now.

why a new PR? maybe just make an extra commit

Reply:
Okay, agreed, I'll do it.